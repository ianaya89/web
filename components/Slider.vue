<template>
  <div class="slider relative">
    <div class="text-p3 text-grey uppercase tracking-wide w-full">
      <slot name="label" />
    </div>
    <template v-if="!range">
      <span class="font-medium text-h5 w-full">{{ formattedCurrentValue }}</span>
    </template>
    <template v-else>
      <span class="font-medium text-h5 w-full">{{ formattedRangeCurrentValue }}</span>
    </template>

    <v-slider
      v-if="!range"
      v-bind="$props"
      class="-mt-1"
      :color="'#FF553F'"
      @change="onChange"
      @input="onInput"
    />
    <v-range-slider
      v-else
      v-bind="$props"
      class="-mt-1"
      :color="'#FF553F'"
      @change="onChange"
      @input="onInput"
    />

    <span class="left-0 bottom-0 absolute text-grey">{{ formattedMin }}</span>
    <span class="right-0 bottom-0 absolute text-grey">{{ formattedMax }}</span>
  </div>
</template>

<script>
export default {
  name: 'Slider',

  props: {
    value: {
      type: [Number, Array],
      required: true
    },

    range: {
      type: Boolean,
      default: false
    },

    formatter: {
      type: Function,
      default: null
    },

    formatterArgs: {
      type: Array,
      default: () => []
    },

    min: {
      type: Number,
      default: 0
    },

    max: {
      type: Number,
      default: 100
    },

    disabled: {
      type: Boolean,
      default: false
    }
  },

  data: () => ({
    currentValue: []
  }),

  computed: {
    formattedCurrentValue() {
      if (this.formatter !== null) {
        return this.formatter(this.currentValue, ...this.formatterArgs)
      } else {
        return this.$filters.number(this.currentValue)
      }
    },

    formattedRangeCurrentValue() {
      if (this.formatter !== null) {
        return (
          this.formatter(this.currentValue[0], ...this.formatterArgs) +
          ' - ' +
          this.formatter(this.currentValue[1], ...this.formatterArgs)
        )
      } else {
        return (
          this.$filters.number(this.currentValue[0]) +
          ' - ' +
          this.$filters.number(this.currentValue[1])
        )
      }
    },

    formattedMin() {
      if (this.formatter !== null) {
        return this.formatter(this.min, ...this.formatterArgs)
      } else {
        return this.$filters.number(this.min)
      }
    },

    formattedMax() {
      if (this.formatter !== null) {
        return this.formatter(this.max, ...this.formatterArgs)
      } else {
        return this.$filters.number(this.max)
      }
    }
  },

  watch: {
    value(val) {
      this.currentValue = val
    }
  },

  mounted() {
    this.currentValue = this.value
  },

  methods: {
    onInput(value) {
      this.currentValue = value
      this.$emit('input', value)
    },
    onChange(value) {
      this.onInput(value)
      this.$emit('change', value)
    }
  }
}
</script>

<style lang="postcss">
.slider .v-slider__thumb {
  @apply .bg-white .border-2 .border-red !important;
}
</style>
