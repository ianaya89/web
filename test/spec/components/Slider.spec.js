import { shallowMount } from '@vue/test-utils'
import Slider from '@/components/Slider'

describe('Slider', () => {
  it('renders default Slider', () => {
    const wrapper = shallowMount(Slider, {
      propsData: { value: 20 }
    })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('renders range Slider specific min max', () => {
    const wrapper = shallowMount(Slider, {
      propsData: {
        range: true,
        value: [20, 500],
        min: 10,
        max: 10000
      }
    })
    expect(wrapper.element).toMatchSnapshot()

    wrapper.setProps({
      disabled: true
    })

    expect(wrapper.element).toMatchSnapshot()
  })
})
