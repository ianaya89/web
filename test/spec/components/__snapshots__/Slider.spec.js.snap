// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider renders default Slider 1`] = `
<div
  class="slider relative"
>
  <div
    class="text-p3 text-grey uppercase tracking-wide w-full"
  />
   
  <span
    class="font-medium text-h5 w-full"
  >
    0
  </span>
   
  <v-slider-stub
    backgroundcolor=""
    class="-mt-1"
    color="#FF553F"
    errorcount="1"
    errormessages=""
    formatterargs=""
    max="100"
    messages=""
    min="0"
    rules=""
    step="1"
    successmessages=""
    thumbsize="32"
    ticklabels=""
    ticksize="1"
    value="20"
  />
   
  <span
    class="left-0 bottom-0 absolute text-grey"
  >
    0
  </span>
   
  <span
    class="right-0 bottom-0 absolute text-grey"
  >
    100
  </span>
</div>
`;

exports[`Slider renders range Slider specific min max 1`] = `
<div
  class="slider relative"
>
  <div
    class="text-p3 text-grey uppercase tracking-wide w-full"
  />
   
  <span
    class="font-medium text-h5 w-full"
  >
    - - -
  </span>
   
  <v-range-slider-stub
    backgroundcolor=""
    class="-mt-1"
    color="#FF553F"
    errorcount="1"
    errormessages=""
    formatterargs=""
    max="10000"
    messages=""
    min="10"
    range="true"
    rules=""
    step="1"
    successmessages=""
    thumbsize="32"
    ticklabels=""
    ticksize="1"
    value="20,500"
  />
   
  <span
    class="left-0 bottom-0 absolute text-grey"
  >
    10
  </span>
   
  <span
    class="right-0 bottom-0 absolute text-grey"
  >
    10 000
  </span>
</div>
`;

exports[`Slider renders range Slider specific min max 2`] = `
<div
  class="slider relative"
>
  <div
    class="text-p3 text-grey uppercase tracking-wide w-full"
  />
   
  <span
    class="font-medium text-h5 w-full"
  >
    - - -
  </span>
   
  <v-range-slider-stub
    backgroundcolor=""
    class="-mt-1"
    color="#FF553F"
    errorcount="1"
    errormessages=""
    formatterargs=""
    max="10000"
    messages=""
    min="10"
    range="true"
    rules=""
    step="1"
    successmessages=""
    thumbsize="32"
    ticklabels=""
    ticksize="1"
    value="20,500"
  />
   
  <span
    class="left-0 bottom-0 absolute text-grey"
  >
    10
  </span>
   
  <span
    class="right-0 bottom-0 absolute text-grey"
  >
    10 000
  </span>
</div>
`;
